packages/vm/src/wasm_backend/gas.rs

```diff
@@ -1,170 +1,23 @@
-use crate::backend::{Api, Querier, Storage};
-use crate::environment::Environment;
-
-/// In Wasmer, the gas limit is set on modules during compilation and is included in the cached modules.
-/// This causes issues when trying to instantiate the same compiled module with a different gas limit.
-/// A fix for this is proposed here: https://github.com/wasmerio/wasmer/pull/996.
-///
-/// To work around this limitation, we set the gas limit of all Wasmer instances to this very high value,
-/// assuming users won't request more than this amount of gas. In order to set the real gas limit, we pretend
-/// to consume the difference between the two in `set_gas_left` ("points used" in the metering middleware).
-/// Since we observed overflow behaviour in the points used, we ensure both MAX_GAS_LIMIT and points used stay
-/// far below u64::MAX.
-// const MAX_GAS_LIMIT: u64 = u64::MAX / 2;
-
-const FAKE_GAS_AVAILABLE: u64 = 1_000_000;
-
-#[derive(Debug)]
-pub struct InsufficientGasLeft;
-
-/// Decreases gas left by the given amount.
-/// If the amount exceeds the available gas, the remaining gas is set to 0 and
-/// an InsufficientGasLeft error is returned.
-pub fn decrease_gas_left<A: Api, S: Storage, Q: Querier>(
-    env: &Environment<A, S, Q>,
-    amount: u64,
-) -> Result<(), InsufficientGasLeft> {
-    let remaining = get_gas_left(env);
-    if amount > remaining {
-        set_gas_left(env, 0);
-        Err(InsufficientGasLeft)
-    } else {
-        set_gas_left(env, remaining - amount);
-        Ok(())
-    }
-}
-
-/// Set the amount of gas units that can be used in the context.
-pub fn set_gas_left<A: Api, S: Storage, Q: Querier>(_env: &Environment<A, S, Q>, _amount: u64) {}
-
-/// Get how many more gas units can be used in the context.
-pub fn get_gas_left<A: Api, S: Storage, Q: Querier>(_env: &Environment<A, S, Q>) -> u64 {
-    FAKE_GAS_AVAILABLE
+use std::convert::TryInto;
+use wasmer::Instance as WasmerInstance;
+
+// TODO: Use get_remaining_points from wasmer_middlewares (https://github.com/wasmerio/wasmer/pull/1941)
+pub fn get_remaining_points(instance: &WasmerInstance) -> u64 {
+    instance
+        .exports
+        .get_global("remaining_points")
```

Comment:
Note: I wonder if it is posible for a malicious wasm blob to modify this directly.

It is added during instrumentation, but if the wasm code can look up a global by name (maybe that is only available to the caller, not the code itself), then maybe a malicious wasm blob could hold a hook to clear it's gas costs in a loop.

Reply:
Nice one. I'll investigate.

Reply:
I got response from the Wasmer team:

> [..] The index of a global in global.get or global.set is a constant and verified by the wasm parser, not a value that can be computed at runtime. So when we add one more [global], none of the existing code could be accessing it. There also isn't some other way to access it, like you can't form a pointer to a wasm global in the guest because wasm globals don't live in the wasm memory.

Reply:
Very nice. Thanks for following up on this one. And it makes me rest easier that they have considered that attack.